dsa/Algorithms/kadanes-algorithm.md

```diff
@@ -0,0 +1,200 @@
+    - max_so_far = INT_MIN
+    - max_ending_here = 0
+
+  ##### Loop for each element of the array
+
+  - (a) max_ending_here = max_ending_here + a[i]
+  - (b) if(max_so_far < max_ending_here)
+            max_so_far = max_ending_here
+  - (c) if(max_ending_here < 0)
+            max_ending_here = 0
+return max_so_far
+
+
+## Implementing Kadane’s Algorithm
+
+### Python Implementation
+
+```python
+import sys
+
+def maxSubarraySum(arr, n):
+    maxi = -sys.maxsize-1 # maximum sum
+    sum = 0
+
+    for i in range(n):
+        sum += arr[i]
+
+        if sum > maxi:
+            maxi = sum
+
+        # If sum < 0: discard the sum calculated
+        if sum < 0:
+            sum = 0
+
+    # To consider the sum of the empty subarray
+    # uncomment the following check:
+
+    #if maxi < 0: maxi = 0
+
+    return maxi
+
+```
+
+### Java Implementation
+
+```java
+
+import java.util.*;
+
+public class Main {
+    public static long maxSubarraySum(int[] arr, int n) {
+        long maxi = Long.MIN_VALUE; // maximum sum
+        long sum = 0;
+
+        for (int i = 0; i < n; i++) {
+
+            sum += arr[i];
+
+            if (sum > maxi) {
+                maxi = sum;
+            }
+
+            // If sum < 0: discard the sum calculated
+            if (sum < 0) {
+                sum = 0;
+            }
+        }
+
+        // To consider the sum of the empty subarray
+        // uncomment the following check:
+
+        //if (maxi < 0) maxi = 0;
+
+        return maxi;
+    }
+
+    public static void main(String args[]) {
+        int[] arr = { -2, 1, -3, 4, -1, 2, 1, -5, 4};
+        int n = arr.length;
+        long maxSum = maxSubarraySum(arr, n);
+        System.out.println("The maximum subarray sum is: " + maxSum);
+
+    }
+
+}
+```
+
+### C++ Implementation
+
+```cpp
+long long maxSubarraySum(int arr[], int n) {
+    long long maxi = LONG_MIN; // maximum sum
+    long long sum = 0;
+
+    for (int i = 0; i < n; i++) {
+
+        sum += arr[i];
+
+        if (sum > maxi) {
+            maxi = sum;
+        }
+
+        // If sum < 0: discard the sum calculated
+        if (sum < 0) {
+            sum = 0;
+        }
+    }
+
+    // To consider the sum of the empty subarray
+    // uncomment the following check:
+
+    //if (maxi < 0) maxi = 0;
+
+    return maxi;
+}
+
+```
+
+### JavaScript Implementation
+
+```javascript
+function maxSubarraySum(arr, n) {
+    let maxi = Number.MIN_SAFE_INTEGER; // maximum sum
+    let sum = 0;
+
+    for (let i = 0; i < n; i++) {
+        sum += arr[i];
+
+        if (sum > maxi) {
+            maxi = sum;
+        }
+
+        // If sum < 0: discard the sum calculated
+        if (sum < 0) {
+            sum = 0;
+        }
+    }
+
+    // To consider the sum of the empty subarray
+    // uncomment the following check:
+
+    //if (maxi < 0) maxi = 0;
+
+    return maxi;
+}
+```
+
+## Complexity Analysis
+   #### Time Complexity : 
+    - O(n) , We are using a single loop running N times.
```

Comment:
replace `O(n)` to `$O(n)$` and also, `O(1)` to `$O(1)$`

Reply:
Made the Changes. Please Look into it.